incubator/hnc/internal/validators/hncconfig.go

```diff
@@ -141,6 +147,91 @@ func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationS
 	return allow("")
 }
 
+func (c *HNCConfig) checkForest(inst *api.HNCConfiguration) admission.Response {
+	c.Forest.Lock()
+	defer c.Forest.Unlock()
+
+	// Get types that are changed from other modes to "Propagate" mode.
+	gvks := c.getNewPropagateTypes(inst)
+
+	// Check if user-created objects would be overwritten by these mode changes.
+	for gvk, _ := range gvks {
+		conflicts := c.checkConflictsForGVK(gvk)
+		if len(conflicts) != 0 {
+			msg := fmt.Sprintf("Cannot update configuration because setting type %q to 'Propagate' mode would overwrite user-created object(s):\n", gvk)
+			msg += strings.Join(c.checkConflictsForGVK(gvk), "\n")
+			msg += "\nTo fix this, please rename or remove the conflicting objects first."
+			return deny(metav1.StatusReasonConflict, msg)
+		}
+	}
+
+	return allow("")
+}
+
+// ancestorObjects maps an object name to the ancestor namespace(s) in which
+// it's defined.
+type ancestorObjects map[string][]string
+
+// checkConflictsForGVK looks for conflicts from top down for each tree.
+func (c *HNCConfig) checkConflictsForGVK(gvk schema.GroupVersionKind) []string {
+	conflicts := []string{}
+	for _, ns := range c.Forest.GetRoots() {
+		conflicts = append(conflicts, c.checkConflictsForNS(gvk, ancestorObjects{}, ns)...)
+	}
+	return conflicts
+}
+
+func (c *HNCConfig) checkConflictsForNS(gvk schema.GroupVersionKind, ao ancestorObjects, ns *forest.Namespace) []string {
+	conflicts := []string{}
+	objs := ao.copy()
+	for _, o := range ns.GetSourceObjects(gvk) {
+		onm := o.GetName()
+		objs[onm] = append(objs[onm], ns.Name())
+		// If there are more than just the one (current) namespace we just added,
+		// there may be conflicts.
+		if len(objs[onm]) > 1 {
+			// Currently the top conflicting ancestor would overwrite the object.
+			// TODO: check if this is a real conflict
+			conflicts = append(conflicts, fmt.Sprintf("  Object %q in namespace %q would overwrite the one in %q", onm, objs[onm][0], ns.Name()))
```

Comment:
Consider `objs.source(onm)` or `objs.top(onm)` or something like that instead of `objs[onm][0]`?